File: sdks/ruby/test/logging/logger_test.rb
# frozen_string_literal: true

require_relative "../test_helper"

class LoggerTest < Minitest::Test
  def test_level_filtering_info
    logger = Blok::Logging::Logger.new(:info)
    logger.debug("hidden")
    logger.info("visible1")
    logger.warn("visible2")
    logger.error("visible3")

    assert_equal 3, logger.entries.length
  end

  def test_level_filtering_debug
    logger = Blok::Logging::Logger.new(:debug)
    logger.debug("visible")
    logger.info("visible")

    assert_equal 2, logger.entries.length
  end

  def test_level_filtering_error
    logger = Blok::Logging::Logger.new(:error)
    logger.debug("hidden")
    logger.info("hidden")
    logger.warn("hidden")
    logger.error("visible")

    assert_equal 1, logger.entries.length
  end

  def test_entries_contain_correct_levels
    logger = Blok::Logging::Logger.new(:debug)
    logger.debug("d")
    logger.info("i")
    logger.warn("w")
    logger.error("e")

    levels = logger.entries.map(&:level)
    assert_equal %i[debug info warn error], levels
  end

  def test_entries_contain_messages
    logger = Blok::Logging::Logger.new(:debug)
    logger.info("hello world")

    entries = logger.entries
    assert_equal 1, entries.length
    assert_equal "hello world", entries[0].message
  end

  def test_entries_contain_timestamp
    logger = Blok::Logging::Logger.new(:debug)
    logger.info("test")

    entry = logger.entries.first
    assert entry.timestamp
    assert_match(/\d{4}-\d{2}-\d{2}/, entry.timestamp)
  end

  def test_info_with_fields
    logger = Blok::Logging::Logger.new(:debug)
    logger.info("test", fields: { "key" => "value" })

    entry = logger.entries.first
    assert_equal({ "key" => "value" }, entry.fields)
  end

  def test_lines_format
    logger = Blok::Logging::Logger.new(:debug)
    logger.info("hello")
    logger.error("oops")

    lines = logger.lines
    assert_equal 2, lines.length
    assert_match(/\[INFO\]/, lines[0])
    assert_match(/hello/, lines[0])
    assert_match(/\[ERROR\]/, lines[1])
    assert_match(/oops/, lines[1])
  end

  def test_lines_with_fields
    logger = Blok::Logging::Logger.new(:debug)
    logger.info("test", fields: { "x" => 1 })

    line = logger.lines.first
    assert_match(/\[INFO\]/, line)
    assert_match(/"x":1/, line.gsub(" ", ""))
  end

  def test_clear
    logger = Blok::Logging::Logger.new(:debug)
    logger.info("test")
    assert_equal 1, logger.entries.length

    logger.clear
    assert_equal 0, logger.entries.length
  end

  def test_entries_returns_copy
    logger = Blok::Logging::Logger.new(:debug)
    logger.info("test")

    entries = logger.entries
    entries.clear

    # Original should be unaffected
    assert_equal 1, logger.entries.length
  end

  def test_thread_safety
    logger = Blok::Logging::Logger.new(:debug)

    threads = 10.times.map do |i|
      Thread.new { 100.times { logger.info("thread-#{i}") } }
    end
    threads.each(&:join)

    assert_equal 1000, logger.entries.length
  end

  def test_log_entry_to_hash
    entry = Blok::Logging::LogEntry.new(
      level: :info,
      message: "test message",
      fields: { "key" => "val" }
    )

    hash = entry.to_hash
    assert_equal "INFO", hash["level"]
    assert_equal "test message", hash["message"]
    assert hash["timestamp"]
    assert_equal({ "key" => "val" }, hash["fields"])
  end

  def test_log_level_parse
    assert_equal :debug, Blok::Logging::LogLevel.parse("DEBUG")
    assert_equal :info, Blok::Logging::LogLevel.parse("INFO")
    assert_equal :warn, Blok::Logging::LogLevel.parse("WARN")
    assert_equal :error, Blok::Logging::LogLevel.parse("ERROR")
    assert_equal :info, Blok::Logging::LogLevel.parse("unknown")
    assert_equal :debug, Blok::Logging::LogLevel.parse("debug")
  end

  def test_log_level_priority
    assert Blok::Logging::LogLevel.priority(:debug) <
           Blok::Logging::LogLevel.priority(:info)
    assert Blok::Logging::LogLevel.priority(:info) <
           Blok::Logging::LogLevel.priority(:warn)
    assert Blok::Logging::LogLevel.priority(:warn) <
           Blok::Logging::LogLevel.priority(:error)
  end
end

File: sdks/php/tests/Logging/LoggerTest.php
<?php

declare(strict_types=1);

namespace Blok\Nanoservice\Tests\Logging;

use Blok\Nanoservice\Logging\LogEntry;
use Blok\Nanoservice\Logging\Logger;
use Blok\Nanoservice\Logging\LogLevel;
use PHPUnit\Framework\TestCase;

final class LoggerTest extends TestCase
{
    public function testLevelFiltering(): void
    {
        $logger = new Logger(LogLevel::Info);
        $logger->debug('hidden');
        $logger->info('visible');
        $logger->warn('visible');
        $logger->error('visible');

        $this->assertCount(3, $logger->entries());
    }

    public function testDebugLevelShowsAll(): void
    {
        $logger = new Logger(LogLevel::Debug);
        $logger->debug('visible');
        $logger->info('visible');
        $logger->warn('visible');
        $logger->error('visible');

        $this->assertCount(4, $logger->entries());
    }

    public function testErrorLevelShowsOnlyErrors(): void
    {
        $logger = new Logger(LogLevel::Error);
        $logger->debug('hidden');
        $logger->info('hidden');
        $logger->warn('hidden');
        $logger->error('visible');

        $this->assertCount(1, $logger->entries());
        $this->assertSame(LogLevel::Error, $logger->entries()[0]->level);
    }

    public function testLogWithFields(): void
    {
        $logger = new Logger(LogLevel::Debug);
        $logger->infoWith('test', ['key' => 'value']);

        $entries = $logger->entries();
        $this->assertCount(1, $entries);
        $this->assertNotNull($entries[0]->fields);
        $this->assertSame('value', $entries[0]->fields['key']);
    }

    public function testAllLogMethods(): void
    {
        $logger = new Logger(LogLevel::Debug);

        $logger->debug('debug msg');
        $logger->debugWith('debug fields', ['a' => 1]);
        $logger->info('info msg');
        $logger->infoWith('info fields', ['b' => 2]);
        $logger->warn('warn msg');
        $logger->warnWith('warn fields', ['c' => 3]);
        $logger->error('error msg');
        $logger->errorWith('error fields', ['d' => 4]);

        $this->assertCount(8, $logger->entries());
    }

    public function testLines(): void
    {
        $logger = new Logger(LogLevel::Debug);
        $logger->info('hello');
        $logger->error('oops');

        $lines = $logger->lines();
        $this->assertCount(2, $lines);
        $this->assertStringContainsString('[INFO]', $lines[0]);
        $this->assertStringContainsString('hello', $lines[0]);
        $this->assertStringContainsString('[ERROR]', $lines[1]);
        $this->assertStringContainsString('oops', $lines[1]);
    }

    public function testLinesWithFields(): void
    {
        $logger = new Logger(LogLevel::Debug);
        $logger->infoWith('request', ['method' => 'POST']);

        $lines = $logger->lines();
        $this->assertCount(1, $lines);
        $this->assertStringContainsString('POST', $lines[0]);
    }

    public function testClear(): void
    {
        $logger = new Logger(LogLevel::Debug);
        $logger->info('test');
        $this->assertCount(1, $logger->entries());

        $logger->clear();
        $this->assertCount(0, $logger->entries());
    }

    public function testEntryTimestampFormat(): void
    {
        $logger = new Logger(LogLevel::Debug);
        $logger->info('test');

        $entry = $logger->entries()[0];
        // Should be ISO 8601 format
        $this->assertMatchesRegularExpression(
            '/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/',
            $entry->timestamp,
        );
    }

    public function testLogEntryToString(): void
    {
        $entry = new LogEntry(
            level: LogLevel::Info,
            message: 'test message',
            timestamp: '2024-01-01T00:00:00Z',
        );

        $this->assertSame('[INFO] 2024-01-01T00:00:00Z test message', (string) $entry);
    }

    public function testLogEntryToStringWithFields(): void
    {
        $entry = new LogEntry(
            level: LogLevel::Error,
            message: 'failed',
            timestamp: '2024-01-01T00:00:00Z',
            fields: ['code' => 500],
        );

        $str = (string) $entry;
        $this->assertStringContainsString('[ERROR]', $str);
        $this->assertStringContainsString('failed', $str);
        $this->assertStringContainsString('500', $str);
    }

    public function testLogLevelPriority(): void
    {
        // assertLessThan($expected, $actual) asserts $actual < $expected
        $this->assertLessThan(LogLevel::Info->priority(), LogLevel::Debug->priority());
        $this->assertLessThan(LogLevel::Warn->priority(), LogLevel::Info->priority());
        $this->assertLessThan(LogLevel::Error->priority(), LogLevel::Warn->priority());
    }

    public function testLogLevelFromString(): void
    {
        $this->assertSame(LogLevel::Debug, LogLevel::fromString('DEBUG'));
        $this->assertSame(LogLevel::Info, LogLevel::fromString('INFO'));
        $this->assertSame(LogLevel::Warn, LogLevel::fromString('WARN'));
        $this->assertSame(LogLevel::Warn, LogLevel::fromString('WARNING'));
        $this->assertSame(LogLevel::Error, LogLevel::fromString('ERROR'));
        $this->assertSame(LogLevel::Info, LogLevel::fromString('UNKNOWN'));
    }
}
